numpy/_core/src/multiarray/stringdtype/casts.c

```diff
@@ -1503,13 +1507,258 @@ STRING_TO_CFLOAT_CAST(clongdouble, l, longdouble)
+
+static NPY_CASTING
+string_to_void_resolve_descriptors(PyObject *NPY_UNUSED(self),
+                                   PyArray_DTypeMeta *NPY_UNUSED(dtypes[2]),
+                                   PyArray_Descr *given_descrs[2],
+                                   PyArray_Descr *loop_descrs[2],
+                                   npy_intp *NPY_UNUSED(view_offset))
+{
+    if (given_descrs[1] == NULL) {
+        // currently there's no way to determine the correct output
+        // size, so set an error and bail
+        PyErr_SetString(
+                PyExc_TypeError,
+                "Casting from StringDType to a fixed-width dtype with an "
+                "unspecified size is not currently supported, specify "
+                "an explicit size for the output dtype instead.");
+        return (NPY_CASTING)-1;
+    }
+    else {
+        // reject structured voids
+        if (given_descrs[1]->names != NULL || given_descrs[1]->subarray != NULL) {
+            PyErr_SetString(
+                    PyExc_TypeError,
+                    "Casting from StringDType to a structured dtype is not "
+                    "supported.");
+            return (NPY_CASTING)-1;
+        }
+        Py_INCREF(given_descrs[1]);
+        loop_descrs[1] = given_descrs[1];
+    }
+
+    Py_INCREF(given_descrs[0]);
+    loop_descrs[0] = given_descrs[0];
+
+    return NPY_UNSAFE_CASTING;
+}
+
+static int
+string_to_void(PyArrayMethod_Context *context, char *const data[],
+               npy_intp const dimensions[], npy_intp const strides[],
+               NpyAuxData *NPY_UNUSED(auxdata))
+{
+    StringDTypeObject *descr = (StringDTypeObject *)context->descriptors[0];
+    npy_string_allocator *allocator = NpyString_acquire_allocator(descr);
+    int has_null = descr->na_object != NULL;
+    int has_string_na = descr->has_string_na;
+    const npy_static_string *default_string = &descr->default_string;
+    const npy_static_string *na_name = &descr->na_name;
+    npy_intp N = dimensions[0];
+    char *in = data[0];
+    char *out = data[1];
+    npy_intp in_stride = strides[0];
+    npy_intp out_stride = strides[1];
+    size_t max_out_size = context->descriptors[1]->elsize;
+
+    while (N--) {
+        const npy_packed_static_string *ps = (npy_packed_static_string *)in;
+        npy_static_string s = {0, NULL};
+        int is_null = NpyString_load(allocator, ps, &s);
+        if (is_null == -1) {
+            npy_gil_error(PyExc_MemoryError,
+                          "Failed to load string in void to string cast");
+            goto fail;
+        }
+        else if (is_null) {
+            if (has_null && !has_string_na) {
+                // lossy but not much else we can do
+                s = *na_name;
+            }
+            else {
+                s = *default_string;
+            }
+        }
+
+        // This might truncate a UTF-8 character. Should we warn if that
+        // happens?  UTF-8 won't be round-trippable if there is truncation
```

Comment:
I don't care, it's an unsafe cast since we don't know if truncation will happen.  `void` is a weird monster anyway...
So, overall I think this is fine.